### PR TITLE
CCM-5694: added scripts to automate repointing comms api at dynamic environment

### PR DESCRIPTION
**Pull request by DavidFueryBJSS**
## Summary
added scripts to automate repointing comms api at dynamic environment

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Checklist
* [x] Brief description of work completed, and any technical decisions made as part of the PR
* [ ] PR link added as a comment to the relevant JIRA ticket
* [ ] PR link shared on Slack and/or Teams
* [ ] 2 reviews received
* [ ] Tester approval

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -1,2 +1,3 @@
 python 3.11.8
 poetry 1.5.0
+jq 1.6
```
Comment (aaron-downing) [inline .tool-versions:3]: Could you add a readme for the new script for how to use it and what it does please

File: scripts/repoint_frontend.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    # modify AssignMessage.MessageBatches.Create.Request.xml
+    file_path = 'proxies/shared/policies/AssignMessage.MessageBatches.Create.Request.xml'
+    block_to_follow = [
+        '<AssignVariable>',
+        '<Name>requestpath</Name>',
+        '<Value>/api/v1/send</Value>',
+        '</AssignVariable>'
+    ]
+    lines_to_add = [
+        '    {{% if ENVIRONMENT_TYPE != \'sandbox\' %}}',
```
Comment (aaron-downing) [inline scripts/repoint_frontend.py:137]: Could you not encode the line endings and just have this as a string instead of an array to make the file manipulation more straightforward?

File: scripts/repoint_frontend.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        '            <Name>target.url</Name>',
+        '            <Value>https://comms-apim.de-{shortcode}.communications.national.nhs.uk/api/v1/messages/{{data.messageId}}</Value>',  # noqa: E501
+        '        </AssignVariable>',
+        '    {% endif %}'
+    ]
+    add_lines_after_block(args.shortcode, file_path, block_to_follow, lines_to_add)
+
+    # modify proxies/shared/policies/AssignMessage.NhsAppAccounts.Get.Request.xml
+    file_path = "proxies/shared/policies/AssignMessage.NhsAppAccounts.Get.Request.xml"
+    block_to_follow = [
```
Comment (aaron-downing) [inline scripts/repoint_frontend.py:184]: Nit: Similarly these could also be constants, I'd rather declutter `if __name__ == "__main__"` of any const declaration so it's easier to see what is going on in terms of calling to methods

File: scripts/repoint_frontend.sh [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+# FROM_STEP: where to continue execution from, used to skips steps already run
+# 0: remove mTLS (if set)
+# 1: create new branch for proxy modifications
+# 2: modify proxy files
+# 3: stage, commit, and push changes
+# 4: await domainName available status (or timeout)
+
+from_step=0
```
Comment (aaron-downing) [inline scripts/repoint_frontend.sh:108]: Instead of having this part where you skip steps, could you not just check if mTLS is enabled and skip if it isn't. The ability to skip steps seems like an unnecessary complexity?

Comment (aaron-downing) [reply scripts/repoint_frontend.sh:108]: Why do you need it to create a branch for you? Could the tool just not work from a branch

Comment (DavidFueryBJSS) [reply scripts/repoint_frontend.sh:108]: Because I'm just that lazy XD it also automatically commits to the branch and pushes

Comment (aaron-downing) [reply scripts/repoint_frontend.sh:108]: Commiting and pushing seems fine but yeah, the ability of skipping steps etc makes this CLI script seem a bit overcomplicated and probably one for just manual intervention

File: scripts/repoint_frontend.py (hunk 2 above)
Comment (DavidFueryBJSS) [reply scripts/repoint_frontend.py:137]: I think that that would make this code less readable and maintainable

Comment (aaron-downing) [reply scripts/repoint_frontend.py:137]: Just noticed that the line adding code is on one line so no notable difference :)

File: scripts/repoint-frontend/repoint_frontend.sh [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  echo ""  # prevents the previous info being overwritten
+
+  if [[ $domain_name_available -eq $false ]]; then
+    warn "timeout reached before domain name became available - mTLS update is likely still processing"
+  fi
+fi
+
+# final output
+if [[ $commit_made -eq $true ]]; then
+  echo ""
```
Comment (DavidFueryBJSS) [inline scripts/repoint-frontend/repoint_frontend.sh:261]: My fault for not updating these, there's a util function up top called spacer that does just this, you **could** either remove that or update the instances of `echo ""` to use it... but only if you want to because it doesn't really matter

**Review by Ian-Hodges — COMMENTED** [review]
Looks good, but it is missing a step to make it 100% automated.
Step 6 on [API Testing specific backend environment](https://nhsd-confluence.digital.nhs.uk/display/RIS/NHS+Notify+%7C+API+Testing+specific+backend+environment).

**Comment by Ian-Hodges** [comment]
> Looks good, but it is missing a step to make it 100% automated. Step 6 on [API Testing specific backend environment](https://nhsd-confluence.digital.nhs.uk/display/RIS/NHS+Notify+%7C+API+Testing+specific+backend+environment).

I've just seen that it creates a new PR so not so much of an issue. When I've repointed the backend it's generally when I wanted to test directly against it from my local machine, after the pipeline has finished.

File: scripts/repoint-frontend/repoint_frontend.sh [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+commit_made=$false
+if check_run_step "stage, commit, and push changes"; then
+  git add proxies
+  exit_on_failure "filed to stage changes (git add)"
+
+  git commit -m "CCM-${ticket_id}: repointed frontend" --no-verify  # verify breaks on my machine, will fix this in due course
+  exit_on_failure "failed to commit changes (git commit)"
+
+  git push --set-upstream origin "$new_branch"
+  exit_on_failure "failed to push changes (git push)"
```
Comment (aaron-downing) [inline scripts/repoint-frontend/repoint_frontend.sh:223]: Not a bash wizard but I'm unsure how this line would ever hit the exit clause of `exit_on_failure`. I'd presume if line 222 returned a non-zero exit code that this script would just exit out?

Comment (DavidFueryBJSS) [reply scripts/repoint-frontend/repoint_frontend.sh:223]: Bash only automatically exits on non-zero codes if you have `set -e` at the top of the file

File: scripts/repoint-frontend/repoint_frontend.sh [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+  if [[ -z "$shortcode" ]]; then
+    error "missing argument: shortcode" "usage: $0 <ticket ID: 0000> <shortcode: gith1>"
+  fi
+fi
+
+info "initial validation passed"
+
+if check_run_step "remove mTLS (if set)"; then
+  # check if mTLS set on env
+  domain_name="comms-apim.de-$shortcode.communications.national.nhs.uk"
```
Comment (DavidFueryBJSS) [inline scripts/repoint-frontend/repoint_frontend.sh:180]: Won't work for de2 environments

File: scripts/repoint-frontend/repoint_frontend.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    (
+        "proxies/shared/policies/AssignMessage.Messages.GetSingle.Request.xml",
+        "/api/v1/messages/{data.messageId}",
+        "https://comms-apim.de-{shortcode}.communications.national.nhs.uk/api/v1/messages/{{data.messageId}}",
+        True,
+    ),
+    (
+        "proxies/shared/policies/AssignMessage.NhsAppAccounts.Get.Request.xml",
+        "/api/channels/nhsapp/accounts",
+        "https://comms-apim.de-{shortcode}.communications.national.nhs.uk/api/channels/nhsapp/accounts",
```
Comment (DavidFueryBJSS) [inline scripts/repoint-frontend/repoint_frontend.py:36]: Won't work for de2 envs

File: scripts/repoint-frontend/repoint_frontend.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+    parser = argparse.ArgumentParser(
+        description="Script to update proxy configuration files based on the environment argument."
+    )
+    parser.add_argument(
+        "environment", type=str, help="Environment identifier (e.g. 'de-gith1')"
+    )
+    args = parser.parse_args()
+
+    # Process `default` block type configs
```
Comment (adadigital) [inline scripts/repoint-frontend/repoint_frontend.py:154]: Could we get a little more illustrative comments? I'm finding it a little hard to parse what a "config" means in the context of `get_config_by_type` and what `match_block` and `insertion_lines` are all for. For a little stand-alone script like this that's ok, and I'm glad I'm not looking at the same thing that's implemented in bash. I just hoped maybe you could help walk the reader through the intention behind what's happening in plain english comments if it's not super clear from the code on its own?

File: scripts/mtls/disable_mtls.sh [hunk 10]
```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+set -e
+
+help() {
+  echo ""
+  echo "Usage: $0 [environment]"
```
Comment (adadigital) [inline scripts/mtls/disable_mtls.sh:6]: Just pop a message in here to help explain what this tool is for and what it does. I might not remember in a few months time. It would help to include some context about why we have this.

File: scripts/repoint-frontend/repoint_frontend.sh [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+# 4: await domainName available status (or timeout)
+
+# sets list_steps_only to true, this is used by check_run_step
+list_steps() {
+  list_steps_only=$true
+  echo "STEPS:"
+}
+
+help() {
+  echo "Usage: $0 [ticket ID] [environment] [options]"
```
Comment (adadigital) [inline scripts/repoint-frontend/repoint_frontend.sh:93]: This also needs a little explanation and context.

File: scripts/repoint-frontend/README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+## Usage
+
+```bash
+./scripts/repoint-frontend/repoint_frontend.sh <ticket ID> <environment> [options]
+```
+
+e.g.
+```bash
+./scripts/repoint-frontend/repoint_frontend.sh 0000 de-todr3
```
Comment (adadigital) [inline scripts/repoint-frontend/README.md:39]: I'm a little confused. Is the intention that we should run both scripts or is it an either-or situation?

**Review by adadigital — CHANGES_REQUESTED** [review]

File: scripts/repoint-frontend/README.md (hunk 12 above)
Comment (lapenna-bjss) [reply scripts/repoint-frontend/README.md:39]: repoint_frontend.sh is the main entry point—it manages all steps and validation. The update_proxy_files.py script is called within repoint_frontend.sh to handle the line replacements in the proxy files.

File: scripts/repoint-frontend/repoint_frontend.sh [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    return $false
+  fi
+
+  echo ""
+  step "running: $1"
+  return $true
+}
+
+# doesn't validate the positional args if only listing steps
+if [[ $list_steps_only -eq $false ]]; then
```
Comment (aaron-downing) [inline scripts/repoint-frontend/repoint_frontend.sh:135]: Do you need it to list the steps it's doing if it's documented in the readme?

File: scripts/repoint-frontend/repoint_frontend.sh [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+if [[ $list_steps_only -ne $true ]]; then
+  info "initial validation passed"
+fi
+
+if check_run_step "remove mTLS (if set)"; then
+  ./scripts/mtls/disable_mtls.sh $environment
+  exit_on_failure echo "Failed to disable mTLS, exiting..."
+fi
+
+if check_run_step "create new branch for proxy modifications"; then
```
Comment (aaron-downing) [inline scripts/repoint-frontend/repoint_frontend.sh:156]: Can you remove these `check_run_step` conditions? I thought they were there previously for when you only wanted to run X step in the script

Comment (lapenna-bjss) [reply scripts/repoint-frontend/repoint_frontend.sh]: Good point. I have removed the `check_run_step` function.

**Review by aaron-downing — APPROVED** [review]

**Review by THerbertBJSS — APPROVED** [review]

File: scripts/repoint-frontend/README.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+## Steps
+
+The script performs the following steps:
+
+1. Remove mTLS if set, ensuring it is disabled if previously configured
+2. Create a new branch to contain changes to the proxy
+3. Update proxy files with the necessary configuration changes
+4. Stage, commit, and push changes to the remote repository
+
```
Comment (Kennedy1265) [inline scripts/repoint-frontend/README.md:29]: Can we include the manual steps required to finish setting up an environment somewhere? You, as a user need to open a PR and navigate to "apps" > "comms-manager-local" in apigee and add the PR product that gets created to get this to work end to end. Might also be worth specifying that in order to test this environment you will need to target comms-pr-xxx

File: scripts/repoint-frontend/README.md [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+The script simplifies this process by automatically:
+
+1. Pointing the APIs to the correct backend environment
+2. Disabling mTLS where necessary
+
+## Prerequisites
+
+Before running the "Repoint Frontend" script, ensure the following prerequisites are met:
+
+- You must be logged into an AWS account using SSO authentication
```
[the remark this answers is not in the document]
Comment (DavidFueryBJSS) [reply scripts/repoint-frontend/README.md:18]: I did consider making it check the user is logged into the correct account (comms mgr dev) but there might be use in making this accessible to more than one AWS account

File: scripts/repoint-frontend/README.md [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+1. Remove mTLS if set, ensuring it is disabled if previously configured
+2. Create a new branch to contain changes to the proxy
+3. Update proxy files with the necessary configuration changes
+4. Stage, commit, and push changes to the remote repository
+
+## Usage
+
+```bash
+./scripts/repoint-frontend/repoint_frontend.sh <ticket ID> <environment>
```
[the remark this answers is not in the document]
Comment (Kennedy1265) [reply scripts/repoint-frontend/README.md:33]: Its in the help text of the command so feel free to leave as is

**Review by adadigital — APPROVED** [review]
